core/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java

```diff
@@ -142,72 +176,160 @@ public TermsQueryBuilder disableCoord(boolean disableCoord) {
         return this;
     }
 
+    public boolean disableCoord() {
+        return this.disableCoord;
+    }
+
+    private boolean isTermsLookupQuery() {
+        return this.termsLookup != null;
+    }
+
+    public TermsQueryBuilder termsLookup(TermsLookup termsLookup) {
+        this.termsLookup = termsLookup;
+        return this;
+    }
+
+    public TermsLookup termsLookup() {
+        return this.termsLookup;
+    }
+
     /**
      * Sets the index name to lookup the terms from.
      */
     public TermsQueryBuilder lookupIndex(String lookupIndex) {
-        this.lookupIndex = lookupIndex;
+        if (lookupIndex == null) {
+            throw new IllegalArgumentException("Lookup index cannot be set to null");
+        }
+        if (this.termsLookup == null) {
+            this.termsLookup = new TermsLookup();
+        }
+        this.termsLookup.setIndex(lookupIndex);
         return this;
     }
 
+    public String lookupIndex() {
+        return this.termsLookup.getIndex();
+    }
+
     /**
-     * Sets the index type to lookup the terms from.
+     * Sets the type name to lookup the terms from.
      */
     public TermsQueryBuilder lookupType(String lookupType) {
-        this.lookupType = lookupType;
+        if (lookupType == null) {
+            throw new IllegalArgumentException("Lookup type cannot be set to null");
+        }
+        if (this.termsLookup == null) {
+            this.termsLookup = new TermsLookup();
+        }
+        this.termsLookup.setType(lookupType);
         return this;
     }
 
+    public String lookupType() {
+        return this.termsLookup.getType();
```

Comment:
this is NPE when termsLookup is null. I am tempted to just expose a termsLookup() getter rather than all the specific getters. Remember that the specific setters are there only for bw comp, but the getters were not there before, so we don't necessarily have to add them.